tools/toolchains/mbed_toolchain.py

```diff
@@ -733,13 +733,17 @@ def link_program(self, r, tmp_path, name):
         new_path = join(tmp_path, head)
         mkdir(new_path)
 
+        # The output file names are derived from the project name, but this can have spaces in it which
+        # messes-up later processing. Replace any spaces in the derived names with '_'
+        tail = tail.replace(" ", "_")
+
         # Absolute path of the final linked file
         if self.config.has_regions:
-            elf = join(tmp_path, name + '_application.elf')
-            mapfile = join(tmp_path, name + '_application.map')
+            elf = join(tmp_path, tail + '_application.elf')
```

Comment:
I think this has unintended consequences, before the change the output directory was:
- `tmp_path` + `head` + `tail`

And after it is:
- `tmp_path` + `tail`

Which I'm guessing was intentional behaviour due to the lines:
```
        new_path = join(tmp_path, head)
        mkdir(new_path)
```

This is very confusing as `new_path` is then never used again.

I believe changing where `tail` has been used so that it uses `new_path` will retain the existing functionality and fix the space issue:
 
```suggestion
            elf = join(new_path, tail + '_application.elf')
```

Reply:
I've also changed the remaining lines in the same pattern in the following commit.